### PR TITLE
SAA-512 Detect and store interesting events for review

### PR DESCRIPTION
**Pull request by timharrison-moj**
This PR adds:
* Cell move events
* Incentive-level change reviews
* Non-association change events

It checks:
* Whether the prisoner is in a prison which is rolled out for activities
* Whether the prisoner has any currently active allocations
* If so, saves the event for review into the EVENT_REVIEW table.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt [hunk 1]
```diff
@@ -3,32 +3,28 @@ package uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.service.events
...
-    if (rolloutPrisonRepository.findByCode(inboundEvent.prisonCode())?.isActivitiesRolledOut() == true) {
-      when (inboundEvent) {
-        is OffenderReceivedEvent -> receivedEventHandler.handle(inboundEvent)
-        is OffenderReleasedEvent -> releasedEventHandler.handle(inboundEvent)
-        else -> log.warn("Unsupported event ${inboundEvent.javaClass.name}")
-      }
-
-      return
+    when (inboundEvent) {
+      is OffenderReceivedEvent -> receivedEventHandler.handle(inboundEvent)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt:22]: Do we need to mark/capture the received and released events as invents of interest if the handle functions return false?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/EventReview.kt [hunk 2]
```diff
@@ -14,21 +14,21 @@ data class EventReview(
...
+  val eventTime: LocalDateTime? = null,
 
-  val prisonCode: String?,
+  val prisonCode: String? = null,
 
-  val prisonerNumber: String?,
+  val prisonerNumber: String? = null,
 
-  val bookingId: Int?,
+  val bookingId: Int? = 0,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/EventReview.kt:27]: Are you sure you want this to be 0 as the default?

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/EventReview.kt:27]: I meant null.. will change now.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt (hunk 1 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt]: Done now.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandlerTest.kt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  fun `stores a received event when allocations exist`() {
+    val inboundEvent = offenderReceivedFromTemporaryAbsence(pentonvillePrisonCode, "123456")
+    val activeAllocations = listOf(allocation().copy(allocationId = 1, prisonerNumber = "123456"))
+    whenever(allocationRepository.findByPrisonCodeAndPrisonerNumber(pentonvillePrisonCode, "123456"))
+      .doReturn(activeAllocations)
+
+    val result = handler.handle(inboundEvent)
+
+    assertThat(result).isTrue
+    verify(rolloutPrisonRepository, times(1)).findByCode(pentonvillePrisonCode)
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandlerTest.kt:86]: `times(1)` is the default, meh.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  private val prisonApiClient: PrisonApiApplicationClient,
+) : EventHandler<InboundEvent> {
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  override fun handle(event: InboundEvent): Boolean {
+    log.info("Checking for interesting event: $event")
+
+    prisonApiClient.getPrisonerDetails(event.prisonerNumber(), fullInfo = true, extraInfo = true).block()?.let { prisoner ->
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt:36]: Do you need the full and extra info here in the call to the prison api?

Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt:36]: Also not unrelated to this PR there is  potential issue with this endpoint.

This is affecting the release event handler at the moment due to a broken contract between the two services.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt [hunk 5]
```diff
@@ -3,32 +3,36 @@ package uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.service.events
...
 
   fun process(inboundEvent: InboundEvent) {
-    if (rolloutPrisonRepository.findByCode(inboundEvent.prisonCode())?.isActivitiesRolledOut() == true) {
-      when (inboundEvent) {
-        is OffenderReceivedEvent -> receivedEventHandler.handle(inboundEvent)
-        is OffenderReleasedEvent -> releasedEventHandler.handle(inboundEvent)
-        else -> log.warn("Unsupported event ${inboundEvent.javaClass.name}")
+    when (inboundEvent) {
+      is OffenderReceivedEvent -> {
+        if (!receivedEventHandler.handle(inboundEvent)) {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt:23]: We may need to take this further and wrap the calls to the handlers in a runCatching { } or the handlers return a Result<?> and then interrogate.

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt (hunk 4 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt:36]: Its only a test - so to match the mocked function, yes..

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt (hunk 5 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt:23]: Yes - possible to get an exception in handlers.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+          log.info("${prisoner.offenderNo} has no active allocations at ${prisoner.agencyId}")
+        }
+      } else {
+        log.info("${prisoner.agencyId} is not a rolled out prison")
+      }
+    }
+    return false
+  }
+
+  private fun getEventType(event: InboundEvent) =
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt:61]: Yes, good suggestion. Will do it now. Should have done it anyway, but slightly distracted by other stuff.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt (hunk 4 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt:36]: No, don't think I do. Will remove now.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt (hunk 5 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt:23]: Ok to leave this until next part?

Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/InboundEventsService.kt:23]: Yes sorry it was more just a comment.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt (hunk 4 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/InterestingEventHandler.kt:36]: Removed these now.

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**